packages/node_modules/@enocean-js/serialport-parser/src/serialport-parser.js

```diff
@@ -1,111 +1,103 @@
-// Copyright 2015 Rafael Kähm <rafael@kaehms.de>
 // Copyright 2018 Holger Will <h.will@klimapartner.de>
-import getCRC8 from '@enocean-js/crc8'
-// import { ESP3Packet } from '@enocean-js/esp3-packets'
+import { makeESP3Packet } from '@enocean-js/esp3-packets'
 
 const Transform = require('stream').Transform
 
-class ESP3Parser extends Transform {
+const WAIT_FOR_SYNC_BYTE = 0
+const FILL_HEADER = 1
+const CHECK_CRC8_HEADER = 2
+const FILL_DATA_OPTIONALDATA = 3
+const CHECK_CRC8_DATAS = 4
+const MAX_BUFFER_SIZE = 1000
+
+export default class ESP3Parser extends Transform {
   constructor (options = {}) {
     super({ ...options, ...{ readableObjectMode: true } })
-    this.currentESP3Packet = {}
-    this.rawString = ''
-    this.tmp = null
-    this.callbackForNextByte = this.waitForSyncByte
+    this.currentESP3Packet = makeESP3Packet()
+    this.state = WAIT_FOR_SYNC_BYTE
   }
   _transform (chunk, encoding, cb) {
     for (var offset = 0; offset < chunk.length; offset++) {
-      this.rawString += chunk[offset].toString(16).padStart(2, '0')
-      this.callbackForNextByte(chunk[offset], cb)
+      let byte = chunk[offset]
+      switch (this.state) {
+        case WAIT_FOR_SYNC_BYTE:
+          if (byte === 0x55) {
+            this.currentESP3Packet = makeESP3Packet(byte)
+            this.state = FILL_HEADER
+          }
+          break
+        case FILL_HEADER:
+          this.currentESP3Packet.push(byte)
+          if (this.currentESP3Packet.length < 5) {
+            break
+          }
+          this.state = CHECK_CRC8_HEADER
+          break
+        case CHECK_CRC8_HEADER:
+          this.currentESP3Packet.push(byte)
+          if (!this.currentESP3Packet.isHeaderOK()) {
+            var syncCodeIndex = this.currentESP3Packet.findIndex((item, index) => { return (item === 0x55) && (index > 0) })
+            if (syncCodeIndex > 0) {
+              this.state = FILL_HEADER
+              while (syncCodeIndex > 0) {
+                this.currentESP3Packet.shift()
+                syncCodeIndex--
+              }
+              break
+            }
+            this.emit('error', {
+              code: 1,
+              name: 'WRONG_HEADER_CHECKSUM',
+              desc: 'header checksum test failed'
+            })
+            this.state = WAIT_FOR_SYNC_BYTE
+            break
+          }
+          if (this.currentESP3Packet.dataLength + this.currentESP3Packet.optionalLength <= 0) {
+            this.emit('error', {
+              code: 1,
+              name: 'ILLEGAL_PACKET_LENGTH',
+              desc: 'there must be at least 1 byte of data or optional data, it can not be 0'
+            })
+            this.state = WAIT_FOR_SYNC_BYTE
+            break
+          }
+          this.state = FILL_DATA_OPTIONALDATA
+          break
+        case FILL_DATA_OPTIONALDATA:
+          this.currentESP3Packet.push(byte)
+          if (this.currentESP3Packet.length > MAX_BUFFER_SIZE) {
```

Comment:
on the other hand there is `RET_BUFFER_TO_SMALL` -> "The internal ESP3 buffer of the device is too small, to handle this telegram" which suggests that buffer size is application dependent. 

As stupid as it may sound, but i read `"Max. size of transferred data" is 65535 bytes ` as the data field can be 0xffff long because of the dataLength field being 2 Bytes...

In the same table it reads "Upward compatible with ’Optional Data'" so max packet size is 0xffff (dataLenght) + 0xff (optionalLength) + 7 (Sync + header + CRCs) = 65797

So i think we can set it to anything we like. I choose 1000 because its reasonably large. i don't believe there are real world packages that are larger. thought there might be in MSC Land... 

with RADIO_MESSAGE you can send arbitrary messages that are 65535 - 1 Bytes Long + 12 Bytes for the rest of the telegram. 

I'm happy to change it to any number... going with 65535 for now...

Reply:
or what do you think?

Reply:
we could make this an option for the parser.

f.e. If you implement a temperature data logger, you know you will always receive packets no longer than 24 Bytes, why allow for more? this will secure your application and make it faster and use less memory.